### PR TITLE
test/openshift3 with yaml deployment

### PR DESCRIPTION
**Pull request by ivanstanev**
- [x] Tests written and linted [ℹ︎](https://github.com/snyk/general/wiki/Tests)
- [ ] Documentation written [ℹ︎](https://github.com/snyk/general/wiki/Documentation)
- [x] Commit history is tidy [ℹ︎](https://github.com/snyk/general/wiki/Git)
- [ ] Potential release notes have been inspected

### What this does

Use a KinD cluster pinned to Kubernetes version 1.11.10. This allows us to simulate an OpenShift 3 environment.

### More information

- [Jira ticket RUN-826](https://snyksec.atlassian.net/browse/RUN-826)

### CODE REVIEW

File: test/helpers/kubectl.ts [hunk 1]
```diff
@@ -1,21 +1,29 @@
...
 import { platform } from 'os';
 import { resolve } from 'path';
 import * as needle from 'needle';
 import * as sleep from 'sleep-promise';
 
-export async function downloadKubectl(): Promise<void> {
+/**
+ * @param version For example: "v1.18.0"
+ */
+export async function downloadKubectl(version?: string): Promise<void> {
```
Comment (Shesekino) [inline test/helpers/kubectl.ts]: I strongly prefer avoiding optional parameters if possible

File: test/helpers/kubectl.ts [hunk 2]
```diff
@@ -1,21 +1,29 @@
...
+    if (version === undefined) {
+      return;
+    }
+
+    // Always start clean when requesting a specific version.
+    unlinkSync(kubectlPath);
   }
 
-  console.log('Downloading kubectl...');
+  console.log(`Downloading kubectl ${version || ''}...`);
```
Comment (Shesekino) [inline test/helpers/kubectl.ts]: not sure about it, it looks like code written for backwards compatibility instead of a function written in the first place to solve the problem of "download Kubectl"

File: test/helpers/kubectl.ts [hunk 3]
```diff
@@ -72,7 +72,7 @@ export async function applyK8sYaml(pathToYamlDeployment: string): Promise<void>
 
 export async function createPodFromImage(name: string, image: string, namespace: string) {
   console.log(`Letting Kubernetes decide how to manage image ${image} with name ${name}`);
-  await exec(`./kubectl run ${name} --image=${image} -n ${namespace} -- sleep 999999999`);
+  await exec(`./kubectl run ${name} --generator=run-pod/v1 --image=${image} -n ${namespace} -- sleep 999999999`);
```
Comment (Shesekino) [inline test/helpers/kubectl.ts]: so this is backwards compatibility for the older version of kubectl the OS3 tests will use, correct? where the default generator was deployment rather than pod?
I wonder if it's worth mentioning more explicitly somehow, but up to you

File: test/setup/index.ts [hunk 4]
```diff
@@ -92,10 +92,13 @@ export async function deployMonitor(): Promise<string> {
 
     console.log(`platform chosen is ${testPlatform}, createCluster===${createCluster}`);
 
-    await kubectl.downloadKubectl();
+    // TODO: hack, rewrite this
+    const kubernetesVersion = testPlatform === 'openshift3' ? 'v1.11.10' : undefined;
```
Comment (Shesekino) [inline test/setup/index.ts]: agreed, I would want this to come from the module responsible for Openshift/Openshift3

File: test/setup/index.ts [hunk 5]
```diff
@@ -92,10 +92,13 @@ export async function deployMonitor(): Promise<string> {
...
+    await kubectl.downloadKubectl(kubernetesVersion);
```
Comment (Shesekino) [inline test/setup/index.ts]: if we let each platform decide (or not decide) which kubectl version it wants, maybe indicated by "latest" instead of undefined, we would be able to avoid some of the code written to handle the undefined parts, wdyt?
we'd still need to handle "latest" of course but I think it could be clearer

File: test/setup/platforms/index.ts [hunk 6]
```diff
@@ -35,6 +35,16 @@ const eksSetup: IPlatformSetup = {
   setupTester: eks.setupTester,
 };
 
+// Use a kind cluster pinned to a specific Kubernetes version to mimic OS3.
```
Comment (Shesekino) [inline test/setup/platforms/index.ts]: why is that?
I mean, I believe you 😛 but maybe there's a doc somewhere about it? how did you learn this?
might be worth documenting in the code around here

File: test/setup/platforms/kind.ts [hunk 7]
```diff
@@ -11,9 +11,9 @@ export async function setupTester(): Promise<void> {
   await download(osDistro);
 }
 
-export async function createCluster(): Promise<void> {
+export async function createCluster(version?: string): Promise<void> {
```
Comment (Shesekino) [inline test/setup/platforms/kind.ts]: same idea about the ```?```

File: scripts/circleci-jobs/openshift3-integration-tests.py [hunk 8]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/python3
+
+import os
+import subprocess
+
+DOCKERHUB_USER = os.getenv('DOCKERHUB_USER')
+DOCKERHUB_PASSWORD = os.getenv('DOCKERHUB_PASSWORD')
+PROCESS = subprocess.run(["docker", "login", "--username", DOCKERHUB_USER, "--password", DOCKERHUB_PASSWORD], capture_output=True, text=True)
```
Comment (Shesekino) [inline scripts/circleci-jobs/openshift3-integration-tests.py]: I think you can use ```subprocess.get_output``` for that

File: scripts/circleci-jobs/openshift3-integration-tests.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/python3
+
+import os
+import subprocess
+
+DOCKERHUB_USER = os.getenv('DOCKERHUB_USER')
```
Comment (Shesekino) [inline scripts/circleci-jobs/openshift3-integration-tests.py]: I would use capital letters for maybe some of the consts you want to shout about, but not for all vars. wdyt? makes it a bit hard to read

File: scripts/circleci-jobs/openshift3-integration-tests.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+DOCKERHUB_PASSWORD = os.getenv('DOCKERHUB_PASSWORD')
+PROCESS = subprocess.run(["docker", "login", "--username", DOCKERHUB_USER, "--password", DOCKERHUB_PASSWORD], capture_output=True, text=True)
+print(PROCESS.stdout)
+
+CIRCLE_BRANCH = os.getenv("CIRCLE_BRANCH")
+if CIRCLE_BRANCH == "staging":
+    IMAGE_TAG = "staging-candidate"
+else:
+    IMAGE_TAG = "discardable"
```
Comment (Shesekino) [inline scripts/circleci-jobs/openshift3-integration-tests.py]: we're running os3 integration tests for regular branches as well? not just staging? I don't think we do that for all non-KinD integration tests, not sure we want to start doing it

File: scripts/circleci-jobs/openshift3-integration-tests.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+CIRCLE_SHA1 = os.getenv("CIRCLE_SHA1")
+KUBERNETES_MONITOR_IMAGE_NAME_AND_TAG = "snyk/kubernetes-monitor:" + IMAGE_TAG + "-" + CIRCLE_SHA1
+
+PROCESS = subprocess.run(["docker", "pull", KUBERNETES_MONITOR_IMAGE_NAME_AND_TAG], capture_output=True, text=True)
+print(PROCESS.stdout)
+
+PROCESS = subprocess.run(["npm", "run", "test:integration:openshift3"], capture_output=True, text=True, env=dict(os.environ))
```
Comment (Shesekino) [inline scripts/circleci-jobs/openshift3-integration-tests.py]: it looks like this script does two things:
1. sets up the integration test environment by pulling the right image
2. wraps npm test

why not let it just handle the setup, and let the circleCI call npm test as it does for all other flows?

File: scripts/circleci-jobs/openshift3-integration-tests.py (hunk 10 above)
Comment (ivanstanev) [reply scripts/circleci-jobs/openshift3-integration-tests.py]: Agreed but this is a bit of a KinD-based integration test which allows us to run it on PRs. The other platform test suffer from requiring exclusivity on a single cluster, which would slow down PRs if opened in parallel.

File: test/setup/platforms/index.ts (hunk 6 above)
Comment (ivanstanev) [reply test/setup/platforms/index.ts]: I think we decided not to try spinning an OpenShift cluster but instead use a regular Kubernetes cluster with the same version that OS3 ships. So that would be 1.11:  https://docs.openshift.com/container-platform/3.11/release_notes/ocp_3_11_release_notes.html#ocp-311-about-this-release
> This release is based on OKD 3.11, and it uses Kubernetes 1.11.

File: test/setup/index.ts (hunk 5 above)
Comment (ivanstanev) [reply test/setup/index.ts]: I find that I'll have to put another variable in the IPlatform interface to achieve this, do we really want to grow that? Or write some function `getKubernetesVersionForPlatform(testPlatform: string)` in the index.ts of the platforms module which will do a switch-case/if for all platforms?

Comment (Shesekino) [reply test/setup/index.ts]: your call, I just really don't want ```test/setup/index.ts``` to have more and more references to each specific platform. one of the reasons we reverted the original Openshift4 PR was just that.
I really want a clean flow of events with clean abstraction as much as possible.
happy to 🍐 on that if you want

File: test/setup/platforms/index.ts (hunk 6 above)
Comment (Shesekino) [reply test/setup/platforms/index.ts]: not sure I was party to this decision, but all the same - why are we satisfied with that?
that is, when does using KinD give us enough confidence to claim we support OS3?

File: scripts/circleci-jobs/openshift3-integration-tests.py (hunk 10 above)
Comment (Shesekino) [reply scripts/circleci-jobs/openshift3-integration-tests.py]: hmmm okay, let's start with your proposal, but be open minded towards pushing it to run only on merges if at some point we decide it's too much hassle to run on every branch?

File: test/setup/platforms/index.ts (hunk 6 above)
Comment (ivanstanev) [reply test/setup/platforms/index.ts]: Because OS3 is just a bunch of components built on top of vanilla Kubernetes. More specifically, it's lots of Deployments and CustomResources utilising the Kubernetes API. There aren't any modifications to the core Kubernetes underneath it. So any potential differences would happen at runtime (e.g. some component may decide to block insecure Deployments) and are not at the core platform.

We don't do anything special with our Deployment, and we have verified it on both vanilla Kubernetes and on OpenShift 4. Both of these give us already a fair amount of confidence that it would just work on OS3 as well, because OS3 and OS4 are not very different in terms of behaviour. Also OS3 support is mostly maintenance updates/patches so we really don't expect something major changing in the platform!

Comment (Shesekino) [reply test/setup/platforms/index.ts]: excellent, thanks for the explanation :)

File: scripts/circleci-jobs/openshift3-integration-tests.py [hunk 12]
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/python3
+
+import os
```
Comment (shaimendel) [inline scripts/circleci-jobs/openshift3-integration-tests.py]: what is specific here to openshift3? wondering about the file name and why it is used only in this flow

File: .circleci/config.yml [hunk 13]
```diff
@@ -29,6 +29,20 @@ main_branches_filter: &main_branches_filter
...
+  setup_node12:
+    description: Setup Node 12
+    steps:
+      - run:
+          command: |
+            export NVM_DIR="/opt/circleci/.nvm"
+            [ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"
+            nvm install v12
+            npm install
+            echo 'export NVM_DIR="/opt/circleci/.nvm"' >> $BASH_ENV
```
Comment (shaimendel) [inline .circleci/config.yml]: this is instead of aliasing v12 to default?

File: scripts/circleci-jobs/openshift3-integration-tests.py (hunk 12 above)
Comment (ivanstanev) [reply scripts/circleci-jobs/openshift3-integration-tests.py]: Nothing specific, the goal is to reduce our CCI script size. Starting with this PR by not adding more to the existing debt

File: .circleci/config.yml (hunk 13 above)
Comment (ivanstanev) [reply .circleci/config.yml]: Not sure what you mean

File: scripts/circleci-jobs/openshift3-integration-tests.py (hunk 12 above)
Comment (shaimendel) [reply scripts/circleci-jobs/openshift3-integration-tests.py]: so maybe it makes sense to call it in a more general name that hints about the file purpose? anyway, your call

File: .circleci/config.yml (hunk 13 above)
Comment (shaimendel) [reply .circleci/config.yml]: me too

Comment (shaimendel) [reply .circleci/config.yml]: I thought that `nvm alias default 12` might do a similar work

File: test/setup/platforms/index.ts [hunk 14]
```diff
@@ -44,6 +44,10 @@ const openshift4Setup: IPlatformSetup = {
   setupTester: openshift4.setupTester,
 };
 
+export function getKubernetesVersionForPlatform(testPlatform: string): string {
```
Comment (Shesekino) [inline test/setup/platforms/index.ts]: I'm missing something, why are we accepting ```testPlatform``` but not using it?

File: test/setup/platforms/index.ts [hunk 15]
```diff
@@ -45,11 +55,17 @@ const openshift4Setup: IPlatformSetup = {
 };
 
 export function getKubernetesVersionForPlatform(testPlatform: string): string {
-  return 'latest';
+  switch (testPlatform) {
```
Comment (Shesekino) [inline test/setup/platforms/index.ts]: ah, it was put in separate commits 😬 found my answer on the next commit

File: scripts/circleci-jobs/openshift3-integration-tests.py [hunk 16]
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/python3
```
Comment (Shesekino) [inline scripts/circleci-jobs/openshift3-integration-tests.py]: maybe rename the file to ```setup-openshift3-integration-test``` or something like that?
